### PR TITLE
Add azurelinux 3.0 images

### PR DESCRIPTION
**Pull request by sbomer**
Add Azure Linux 3.0 preview images, with the same rootfs setup as the current cbl-mariner-2.0 images, using Ubuntu 16.04. ~We plan to introduce Ubuntu 18.04 rootfs for .NET 9 builds, so the naming scheme includes the rootfs ubuntu version (for example: azurelinux-3.0-cross-amd64-ubuntu-16.04).~

This uses a naming scheme with a -net8.0 suffix. We plan to introduce .NET 9 images with a -net9.0 suffix, that use LLVM 18 and an Ubuntu 16.04 rootfs (if we can find a way to build LLVM 18 there), since we're still aiming to target 16.04 in .NET 9 per https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/995#issuecomment-2036124345:

So image names will look like the following:
- azurelinux-3.0-cross-amd64-net8.0
- azurelinux-3.0-cross-amd64-net9.0

### CODE REVIEW

File: src/azurelinux/3.0/cross/amd64-ubuntu-16.04/Dockerfile
Comment (richlander) [inline src/azurelinux/3.0/cross/amd64-ubuntu-16.04/Dockerfile:37]: LLVM 16 makes sense for .NET 8 since that's what we started with.

18 is already out. We should decide our approach for .NET 9+. Should we definite a peer set of these images that includes LLVM 18? I'm guessing by the time .NET 10 comes around that we'll be looking at LLVM 20.

File: src/azurelinux/3.0/crossdeps-builder/Dockerfile
Comment (richlander) [inline src/azurelinux/3.0/crossdeps-builder/Dockerfile:12]: We should double check with @jkoritzinsky on the CMake version.

File: src/azurelinux/3.0/crossdeps/Dockerfile
Comment (jkoritzinsky) [inline src/azurelinux/3.0/crossdeps/Dockerfile:38]: Azure Linux 3.0 should have node 20 available in the `nodejs` package so we shouldn't have to manually download it.

Comment (richlander) [reply src/azurelinux/3.0/crossdeps/Dockerfile:38]: That'd sure be nice.

File: src/azurelinux/3.0/crossdeps-builder/Dockerfile
Comment (jkoritzinsky) [reply src/azurelinux/3.0/crossdeps-builder/Dockerfile:12]: This cmake version should be all good (3.28.2)

File: src/azurelinux/3.0/crossdeps/Dockerfile
Comment (richlander) [reply src/azurelinux/3.0/crossdeps/Dockerfile:38]: I see this version: `20.10.0-1.azl3`

**Comment by richlander** [comment]
> We plan to introduce Ubuntu 18.04

We should hold off on that one. That _was_ the original thinking. I think we've concluded that we'll stay the course on 16.04 for at least another release, if not two. We got feedback that made us pause on this one.

@jkotas

File: src/azurelinux/3.0/crossdeps/Dockerfile
Comment (jkoritzinsky) [reply src/azurelinux/3.0/crossdeps/Dockerfile:38]: That should be fine. We only needed node20 for some AzDO tasks and for aspnetcore (in the VMR build).

Comment (richlander) [reply src/azurelinux/3.0/crossdeps/Dockerfile:38]: This is where we should start to partition stuff.

Does ASP.NET Core need to use these images? These images should be EXCLUSIVELY used runtime and other teams who have a native build (with overlapping requirements). We need to break teams using these images who have managed only builds and not accept any of their requests for changes.

I'm going to make a managed only image for teams. It isn't entirely obvious what should be in there yet.

Does `runtime` need those AzDo tasks or is that for other repos/teams?

Comment (jkoritzinsky) [reply src/azurelinux/3.0/crossdeps/Dockerfile:38]: I think runtime needs it for Component Governance, but I'm not positive.

aspnetcore wouldn't be using this image except for in the VMR. We could have separate images for the VMR builds that include the dependencies for all constituent projects if we wanted to keep the separation.

Comment (richlander) [reply src/azurelinux/3.0/crossdeps/Dockerfile:38]: Ah yes, VMR. Sven and I talked about that. OK, we probably need it then. Reality is often annoying.

My point about moving repo (non-VMR) builds to another (MUCH) smaller image still stand.

Comment (jkotas) [reply src/azurelinux/3.0/crossdeps/Dockerfile:38]: Does aspnetcore run any official builds on non-Windows platforms?

> Reality is often annoying.

Right. The current high-level breakdown should be like this:
- If the repo is managed-only, it should not need to run official builds outside Windows at all. Our builds are Windows centric for historic reasons. An exception are source build sanity checks that run on [ubuntu image in CI and 1es-mariner-2 image in internal builds](https://github.com/dotnet/arcade/blob/ce6cb9b392e6a534ffab4e8d9d2dbadcc5491b5a/eng/common/templates-official/job/source-build.yml#L49-L56).
- If the repo builds native components (are there any repos like that outside runtime?), it needs more or less the same image as runtime for official or CI build.

**Comment by mthalman** [comment]
@lbussell - Be aware of the new pipelines in this PR that will need to be wired up.

File: src/azurelinux/3.0/crossdeps/Dockerfile
Comment (jkoritzinsky) [reply src/azurelinux/3.0/crossdeps/Dockerfile:38]: With the VMR, we'll now be running official builds for each constituent repo on each platform (as part of the VMR's build command).

Outside the VMR, I believe ASP.NET Core runs the builds that produce the non-Windows runtime packs on those non-Windows platforms. I believe they do so as it matches the innerloop dev experience well and then they don't have to build custom infrastructure for "build on Windows for non-Windows" into their repo (as they do have some Windows-only assets)

**Comment by lbussell** [comment]
> @lbussell - Be aware of the new pipelines in this PR that will need to be wired up.

@mthalman, it seems like the PR validation did run on this PR, so we should merge this and then set up the public and internal pipelines afterwards, right? So that the file is present to create the pipeline against.

File: src/azurelinux/3.0/cross/amd64-ubuntu-16.04/Dockerfile
Comment (jkoritzinsky) [reply src/azurelinux/3.0/cross/amd64-ubuntu-16.04/Dockerfile:37]: I'd prefer -llvmXX over -netXX as it's more descriptive of what's contained in the package.

**Comment by mthalman** [comment]
> > @lbussell - Be aware of the new pipelines in this PR that will need to be wired up.
> 
> @mthalman, it seems like the PR validation did run on this PR, so we should merge this and then set up the public and internal pipelines afterwards, right? So that the file is present to create the pipeline against.

Yep, exactly

File: src/azurelinux/3.0/cross/amd64-ubuntu-16.04/Dockerfile
Comment (jkotas) [reply src/azurelinux/3.0/cross/amd64-ubuntu-16.04/Dockerfile]: > I'd prefer -llvmXX over -netXX as it's more descriptive of what's contained in the package.

It is impossible to fully describe what's in the image in the name. I expect that we would need the keep tweaking the naming scheme as new requirements show up.

`-netXX` naming is more durable. It simply means "image used to build .NET XX release" whatever that means.

Comment (jkoritzinsky) [reply src/azurelinux/3.0/cross/amd64-ubuntu-16.04/Dockerfile]: The `-netXX` scheme is fine with me.

Comment (richlander) [reply src/azurelinux/3.0/cross/amd64-ubuntu-16.04/Dockerfile]: The advantage of that scheme is that it gives people a hint that a `net80` image probably should NOT still be in use in a .NET 9 or 10 build. Or ... I'm hopeful of that.

File: src/azurelinux/3.0/crossdeps/Dockerfile
Comment (richlander) [reply src/azurelinux/3.0/crossdeps/Dockerfile]: I started this "spec" to get us going on a plan. There are a TON of patterns in pipelines today. We need to reduce those a fair bit. Some are quite specialized, like in `dotnet/installer` but others seem needlessly DIY.

https://github.com/dotnet/dotnet-buildtools-prereqs-docker/issues/997

File: manifest.json
Comment (mthalman) [inline manifest.json:5]: Move this down one line for alphabetical order.

**Review by mthalman — APPROVED** [review]

**Pull request merged**
